### ISSUE TITLE
Roundtripping error for unicode ndarrays with char.encode and char.decode with UTF-32LE and UTF-16LE encodings but not their big endian versions

### ISSUE THREAD

**Issue by frejanordsiek** [numpy/numpy#7594]
``numpy.char.encode`` and ``numpy.char.decode`` should effectively be inverses of each other if given the same encoding. However, I have found this not to be the case. It seems that truncating nulls (``'\x00'``) from the end of the individual elements of bytes ndarrays when accessing individual elements is causing the problem.

For reference, I am working with numpy 1.11.0 and Python 3.4.2 and 2.7.9 on a 64-bit (64-bit python environments as well) little endian machine (Intel i5) running Debian. Though, I think that the same result might happen on a big endian machine, but I lack a machine or virtual machine to test it (my attempts at getting such a virtual machine running have not worked).

If I take a unicode ndarray, run it through ``numpy.char.encode`` with UTF-32BE or UTF-16BE encoding and then run the result through ``numpy.char.decode`` with the same encoding, I get the original array back. This is the following code in Python 2.7 and Python 3.4. The example here is using UTF-16BE

``` python
import numpy as np

encoding = 'UTF-16BE'

a = np.array([[u'abc', u'012', u'ABC\U00010437'], [u'W', u'Q', u'Z']], dtype='U4')
b = np.char.encode(a, encoding)

np.char.decode(b, encoding)
```

It works, giving the following output

Python 3.4
``` python
array([['abc', '012', 'ABC𐐷'],
       ['W', 'Q', 'Z']], 
      dtype='<U4')
```

Python 2.7
``` python
array([[u'abc', u'012', u'ABC\U00010437'],
       [u'W', u'Q', u'Z']], 
      dtype='<U4')
```

But, if I change the encoding to little endian (``'UTF-32LE'` or ``'UTF-16LE'``), I get the following errors instead

Python 3.4
```
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-1-b92715410044> in <module>()
      6 b = np.char.encode(a, codec)
      7 
----> 8 np.char.decode(b, codec)

/home/user/.local/lib/python3.4/site-packages/numpy/core/defchararray.py in decode(a, encoding, errors)
    503     """
    504     return _to_string_or_unicode_array(
--> 505         _vec_string(a, object_, 'decode', _clean_args(encoding, errors)))
    506 
    507 

/usr/lib/python3.4/encodings/utf_16_le.py in decode(input, errors)
     14 
     15 def decode(input, errors='strict'):
---> 16     return codecs.utf_16_le_decode(input, errors, True)
     17 
     18 class IncrementalEncoder(codecs.IncrementalEncoder):

UnicodeDecodeError: 'utf-16-le' codec can't decode byte 0x63 in position 4: truncated data
```

Python 2.7
```
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-1-b92715410044> in <module>()
      6 b = np.char.encode(a, codec)
      7 
----> 8 np.char.decode(b, codec)

/home/user/.local/lib/python2.7/site-packages/numpy/core/defchararray.pyc in decode(a, encoding, errors)
    503     """
    504     return _to_string_or_unicode_array(
--> 505         _vec_string(a, object_, 'decode', _clean_args(encoding, errors)))
    506 
    507 

/usr/lib/python2.7/encodings/utf_16_le.pyc in decode(input, errors)
     14 
     15 def decode(input, errors='strict'):
---> 16     return codecs.utf_16_le_decode(input, errors, True)
     17 
     18 class IncrementalEncoder(codecs.IncrementalEncoder):

UnicodeDecodeError: 'utf16' codec can't decode byte 0x63 in position 4: truncated data
```

If I use the little endian encoding (``'UTF-32LE'`` or ``'UTF-16LE'``) and look at ``a`` and ``b`` closer, the reason for the error becomes more apparent. The dtypes of ``a`` and ``b`` are ``'<U4'`` and ``'S10'`` respectively. But if I look at the dtypes of element ``[1, 0]``, I get ``'<U1'`` and ``'S1'`` respectively.

The trailing null characters (``'\x00'``) are dropped when accessing the individual elements. When trying decode from ``'UTF-16LE'`` or ``'UTF-16LE'``, ``numpy.char.decode`` is grabbing data with an odd number of bytes (dropping the required trailing nulls) and trying to decode it, which can't work because the encoding works with quads or pairs of bytes (thus a number divisible by 4 or 2).

This indicates that neither ``numpy.char.decode`` nor the decode function that ``numpy.char.decode`` passes off to elementwise is padding the ends of the individual string elements with the required null bytes.

This only shows up in little endian encodings since that results in trailing nulls. Big endian encodings don't have that problem.

**Comment by frejanordsiek** [comment]
I have developed a work around. Essentially, to decode back after doing a little endian encoding, one obtains a uint16/32 view, byteswaps it, obtains a view of that using the dtype of the encoded array, and then decodes it with the big endian version of the encoding.

```python
import numpy as np

encoding = 'UTF-16LE'

a = np.array([[u'abc', u'012', u'ABC\U00010437'], [u'W', u'Q', u'Z']], dtype='U4')
b = np.char.encode(a, encoding)

c = b.view(np.uint16).byteswap().view(b.dtype)

d = np.char.decode(c, 'UTF-16BE')
```

If you run ``a == d``, one gets

```
array([[ True,  True,  True],
       [ True,  True,  True]], dtype=bool)
```

**Comment by seberg** [comment]
I think byteswapping for unicode is simply not done. Which is the problem. Problably needs implementation in copyswap function(s) and/or the buffer transfers. Somewgere I opened an issue for that as well, but not sure where. If this is important to you, we could discuss it (I liver doiwn the corrisdor ;)).

**Comment by frejanordsiek** [comment]
Well, it isn't that important to me. I have my work around and even if this issue is fixed, I still have to use the work around instead since I have to support older versions of Numpy in one of my packages.

However, Numpy isn't just used by me. So while I don't need a fix, someone else might, whether now or down the line. Just being a responsible user of this package is all - giving back a bit, by reporting bugs I find. Hmm, reminds me, there was an obscure Numpy bug I found a long time ago and found a work around for but forgot to make an issue for it on here. I should go dig it up.

**Comment by seberg** [comment]
The unicode transfers are really in complete mess. I tried to look at it a bit, but not sure where to start, and it seems not quite trivial to cover everything (and even simple stuff just segfaults for me, making it annoying to track down what is all wrong).

1. We need a new unicode transfer (with swap) function. That seems reasonably easy on first try, if you use the AUX_DATA to just put in an adequately sized inter with swap functions to be called in an inner loop.
2. I do not trust the `copyswapn` function for non-contiguous cases.
3. The padding/truncating versions (when you cast from a smaller string to a longer one and vice versa), also need this type of trick or so.

I am not sure that is actually related to this issue, the decoding stuff is probably broken on its own.

**Comment by charris** [comment]
Yes indeed. Last I worked on that stuff was a couple of years ago and it was messy then.

**Comment by njsmith** [comment]
I don't think this is just a bug in the Unicode functions. We could fix those, but you still have the intrinsic problem that the numpy S dtype can't distinguish between trailing NULs that are part of the string being represented, versus ones that are added as padding for short strings.

**Comment by seberg** [comment]
That is a different issue I think. One is a bug, the other is a shortcoming of our string types that is in some sense there by design.

**Comment by frejanordsiek** [comment]
This issue would likely come up with any other multi-byte encodings, not just UTF-16 and UTF-32. Hard to have a table of all the multi-byte codes in Numpy itself since Python can just keep adding new ones.

A simpler way might just be to detect when a decoding doesn't work due to truncated data, then a null byte to the ``bytes`` object, and if it doesn't work again, add another, and another till at most three are added.

The error message itself is

``'UnicodeDecodeError: 'utf16' codec can't decode byte 0x63 in position 4: truncated data'``

which ends in ``'truncated data'`` so it should be straightforward to detect and then start adding null bytes to the extracted ``bytes`` string that is passed to the ``encodings`` module.

Such a solution would be a bit of a hack job and not the most efficient, but it would possibly work.

**Comment by seberg** [comment]
Yeah, I am sorry, I think I have been mixing up two different issues a lot here. Should clean them up.

**Comment by seberg** [comment]
OK, @njsmith is of course right, it is in some sense a design issue, since we don't know if the string is supposed to be long enough.
The situation would probably be better though, if we just assumed that the zero padding is suitable for the new encoding as well. Fixing this should then not be very hard, may either require some added functionality to the _vec_string function, or simply write something in plain python, and hope the speed loss is not astronomical (which seems likely).

**Comment by danielhrisca** [comment]
Hello,

is there any progress with this issue?

@frejanordsiek 
thank you for the workaround, it works well!

**Comment by seberg** [comment]
I am not aware of anyone working on it. a real fix would probably be some day with new string dtypes....

**Comment by danielhrisca** [comment]
Would it make sense to include this as a note in the _numpy.core.defchararray.decode_ documentation?

**Comment by frejanordsiek** [comment]
I would say that it is definitely worth pointing out both there and with encode the issue with the dropping of trailing nulls.

**Comment by mattip** [comment]
@frejanordsiek could you suggest a sentence or two? A [pull request](http://www.numpy.org/devdocs/dev/gitwash/development_workflow.html) against [the docs](https://github.com/numpy/numpy/blob/master/numpy/core/defchararray.py#L462) would be great, but even a draft something here would be nice
